File: Solution/1041.困于环中的机器人/1041.困于环中的机器人.cpp
/*
 * @Descroption: LeetCode 1041. 困于环中的机器人
 * @Author: EmoryHuang
 * @Date: 2021-03-24 20:03:36
 * @解题思路:
 * 建立四个方向的表示方法，遍历字符串进行运动
 * 只有 `LR`，重复四次之后回到 `[0,0]`，因此需要遍历 4 次
 */

class Solution {
   public:
    bool isRobotBounded(string instructions) {
        int X[4] = {0, 1, 0, -1};
        int Y[4] = {1, 0, -1, 0};
        int flag = 0;
        int x = 0, y = 0;
        for (int j = 0; j < 4; j++) {
            for (int i = 0; i < instructions.size(); i++) {
                if (instructions[i] == 'G') {
                    x += X[flag];  // 按指定方向前进
                    y += Y[flag];
                } else if (instructions[i] == 'L')
                    flag = (flag + 1) % 4;  // 左转
                else
                    flag = (flag + 3) % 4;  // 右转
            }
        }
        return x == 0 && y == 0;
    }
};

File: Solution/1041.困于环中的机器人/1041.困于环中的机器人.py
'''
Descroption: LeetCode 1041. 困于环中的机器人
Author: EmoryHuang
Date: 2023-04-11 09:29:47
Method:
模拟
一轮后不朝北或者回到原点都会陷入循环
'''


class Solution:
    def isRobotBounded(self, instructions: str) -> bool:
        X = [0, 1, 0, -1]
        Y = [1, 0, -1, 0]
        flag = 0
        x, y = 0, 0
        for v in instructions:
            if v == 'G':
                x += X[flag]
                y += Y[flag]
            elif v == 'L':
                flag = (flag - 1) % 4
            elif v == 'R':
                flag = (flag + 1) % 4
        return flag != 0 or (x == 0 and y == 0)
